dbcon/joblist/tuple-bps.cpp

```diff
@@ -2157,256 +2332,93 @@ void TupleBPS::receiveMultiPrimitiveMessages(uint32_t threadID)
 
             tplLock.unlock();
 
-            for (i = 0; i < size && !cancelled(); i++)
+            vector<vector<_CPInfo>> cpInfos;
+            // Calculate the work sizes.
+            uint32_t currentThreadPoolSize = jobstepThreadPool.getThreadCount();
```

Comment:
Could you change a logic a bit here. Now the code takes into account a number of threads in the thread pool but we need to take into account a number of threads free in the pool, namely  max(ThreadPool::fMaxThreads - ThreadPool::fIssued (running functors) , 1). 
It would be good to take waitingFunctorsSize - fIssued(waiting functors) into account also b/c there might be cases when ThreadPool::fIssued is small but number of waiting functors is large. But we can just ignore this corner case now.